# Unit 2
---


https://datacarpentry.org/python-socialsci/08-Pandas/index.html

1. Introducting Pandas
2. Reading files
3. Selecting data
4. Conditional selection




## Pandas
---
<div>
<img src="pandas.JPG" width="400"/>
</div>

[Panda's documentation](https://pandas.pydata.org/pandas-docs/stable/)





To begin we need to import pandas
When you see pd, know it is referring to pandas

In [79]:
import numpy as np
import pandas as pd

## Reading files

<div>
<img src="reading.PNG" width="400"/>
</div>


We will read the whole file at once using Pandas.
Sometimes you might want to read the file line by line, and process each line. Thats possible of course. See for example [here.](https://www.geeksforgeeks.org/read-a-file-line-by-line-in-python/)

Read wisdom of the crowds experiment - ~500,000 answers from ~1700 users to ~1000 questions.
This experiment was performed by researchers in Standford in 2019.[This is the paper they published.](https://ojs.aaai.org//index.php/HCOMP/article/view/5271)  (click on the PDF to download).

[This is their data.](https://github.com/stanford-policylab/wisdom-of-crowds)

Reference to paper:
Simoiu, C., Sumanth, C., Mysore, A., & Goel, S. (2019, October). Studying the “wisdom of crowds” at scale. In Proceedings of the AAAI Conference on Human Computation and Crowdsourcing (Vol. 7, No. 1, pp. 171-179).

In [154]:
import os
cwd = os.getcwd()
cwd

'C:\\Users\\lihi\\OneDrive - ariel.ac.il\\Ariel_Teachings\\a_intro to data analysis\\my lectures\\lecture 2 - data'

read_csv has about 30 different options. See the [documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html#pandas.read_csv)

For example, sep='\t' is used for tab delimited files and 'usecol' reads only specific columns. 

In [18]:
data = pd.read_csv('answers.csv')

view the shape of the dataframe:

In [19]:
data.shape

(516066, 7)

view the column names:

In [20]:
for name in data.columns:
    print(name)

task_id
user_id
answer
confidence
start_time
end_time
cues


View the first few rows:

In [14]:
data.head()

task_id  user_id    answer  confidence           start_time  \
0     1204     2826      left           2  2017-02-02 23:57:06   
1     1910     2827        20           3  2017-02-02 23:57:11   
2     2810     2828  georgian           1  2017-02-02 23:57:17   
3     1208     2826     right           1  2017-02-02 23:57:18   
4     1902     2827        75           2  2017-02-02 23:57:23   

              end_time               cues  
0  2017-02-02 23:57:17  "Not enough data"  
1  2017-02-02 23:57:22  "Not enough data"  
2  2017-02-02 23:57:32  "Not enough data"  
3  2017-02-02 23:57:45  "Not enough data"  
4  2017-02-02 23:57:30  "Not enough data"

`What do you think that the 'tail' command does? Try it out!`

`What happens if we just type data, without a head or a tail?` 

In [15]:
data.head()

task_id  user_id    answer  confidence           start_time  \
0     1204     2826      left           2  2017-02-02 23:57:06   
1     1910     2827        20           3  2017-02-02 23:57:11   
2     2810     2828  georgian           1  2017-02-02 23:57:17   
3     1208     2826     right           1  2017-02-02 23:57:18   
4     1902     2827        75           2  2017-02-02 23:57:23   

              end_time               cues  
0  2017-02-02 23:57:17  "Not enough data"  
1  2017-02-02 23:57:22  "Not enough data"  
2  2017-02-02 23:57:32  "Not enough data"  
3  2017-02-02 23:57:45  "Not enough data"  
4  2017-02-02 23:57:30  "Not enough data"

---
## Selecting data

Select specific rows and/or columns using iloc (index based location):
(note that only the last row will show)

In [27]:
# Rows:
data.iloc[0] # first row of data frame  
data.iloc[1] # second row of data frame  
data.iloc[-1] # last row of data frame  
# Columns:
data.iloc[:,0] # first column of data frame  
data.iloc[:,1] # second column of data frame  
data.iloc[:,-1] # last column of data frame  

task_id  user_id
0          1204     2826
1          1910     2827
2          2810     2828
3          1208     2826
4          1902     2827
...         ...      ...
516061    10574     4708
516062    10571     4708
516063    10579     4708
516064    10575     4708
516065    10573     4708

[516066 rows x 2 columns]

The : operator 

 - when used alone it means "everything"

- also used to indicate a range of values


In [53]:
#Rows and columns
data.iloc[0:5] # first five rows of dataframe
data.iloc[:, 0:2] # first two columns of data frame with all rows
data.iloc[[0,3,6,24], [0,5,6]] # 1st, 4th, 7th, 25th row + 1st 6th 7th columns.

task_id             end_time               cues
0      1204  2017-02-02 23:57:17  "Not enough data"
3      1208  2017-02-02 23:57:45  "Not enough data"
6      1901  2017-02-02 23:57:38  "Not enough data"
24    10064  2017-02-02 23:58:06  "Not enough data"

---
---

Now you: 

what do you do to select:

a. first five rows?

b. first two columns, all rows?

c. 1st and 3rd row and 2nd and 4th column?

---
---

Select by column names using loc (label based location):

In [51]:
data.loc[:,['task_id', 'user_id']]

task_id  user_id
0          1204     2826
1          1910     2827
2          2810     2828
3          1208     2826
4          1902     2827
...         ...      ...
516061    10574     4708
516062    10571     4708
516063    10579     4708
516064    10575     4708
516065    10573     4708

[516066 rows x 2 columns]

In [55]:
data.loc[1:4,['task_id']]

task_id
1     1910
2     2810
3     1208
4     1902

Semantics are similar to iloc. But note:

- `iloc` excludes the last element.  `df.iloc[0:1000]` will return entries 0...999
- `loc`, include the last element. 

you try it! What is the difference between:

> data.iloc[0:5]

> data.loc[0:5]

---
---
Now you:

a. Select the user_id column and assign it into a varialble named: user

b. Use type(user) to check if user is a dataframe or a series

---
---




---
Note: indexing operators as the ones working on dictionaries, will also work in pandas. But for more advanced operations, better get used to loc and iloc.

---

## Conditional selection

Let's select task #1901:

In [63]:
data.loc[:,'task_id'] == 1901

0         False
1         False
2         False
3         False
4         False
          ...  
516061    False
516062    False
516063    False
516064    False
516065    False
Name: task_id, Length: 516066, dtype: bool

This creates a series of true/false 

We can insert this into data to select only that task:

In [64]:
data[data.loc[:,'task_id'] == 1901]

task_id  user_id answer  confidence           start_time  \
6          1901     2827    150           4  2017-02-02 23:57:30   
2942       1901     2825    100           1  2017-02-03 03:32:16   
3358       1901     2925    270           4  2017-02-03 03:42:18   
3764       1901     2841     90           2  2017-02-03 03:51:52   
4548       1901     2972    250           3  2017-02-03 19:44:55   
...         ...      ...    ...         ...                  ...   
510135     1901     4514     80           4  2017-02-24 20:24:14   
511715     1901     4728    120           3  2017-02-24 21:00:07   
512430     1901     4501     80           2  2017-02-24 21:08:50   
513729     1901     4416    110           4  2017-02-24 21:35:40   
515615     1901     3011    150           1  2017-02-24 22:21:25   

                   end_time                   cues  
6       2017-02-02 23:57:38      "Not enough data"  
2942    2017-02-03 03:32:21                    NaN  
3358    2017-02-03 03:42:28                    NaN  
3764    2017-02-03 03:51:58                    NaN  
4548    2017-02-03 19:45:01      "Not enough data"  
...                     ...                    ...  
510135  2017-02-24 20:24:28  ["150", "166", "150"]  
511715  2017-02-24 21:00:17                    110  
512430  2017-02-24 21:08:58     ["160", "80", "2"]  
513729  2017-02-24 21:35:45                    110  
515615  2017-02-24 22:21:30    ["80", "160", "80"]  

[519 rows x 7 columns]

Another way:

In [66]:
data.loc[data.task_id == 1901]

task_id  user_id answer  confidence           start_time  \
6          1901     2827    150           4  2017-02-02 23:57:30   
2942       1901     2825    100           1  2017-02-03 03:32:16   
3358       1901     2925    270           4  2017-02-03 03:42:18   
3764       1901     2841     90           2  2017-02-03 03:51:52   
4548       1901     2972    250           3  2017-02-03 19:44:55   
...         ...      ...    ...         ...                  ...   
510135     1901     4514     80           4  2017-02-24 20:24:14   
511715     1901     4728    120           3  2017-02-24 21:00:07   
512430     1901     4501     80           2  2017-02-24 21:08:50   
513729     1901     4416    110           4  2017-02-24 21:35:40   
515615     1901     3011    150           1  2017-02-24 22:21:25   

                   end_time                   cues  
6       2017-02-02 23:57:38      "Not enough data"  
2942    2017-02-03 03:32:21                    NaN  
3358    2017-02-03 03:42:28                    NaN  
3764    2017-02-03 03:51:58                    NaN  
4548    2017-02-03 19:45:01      "Not enough data"  
...                     ...                    ...  
510135  2017-02-24 20:24:28  ["150", "166", "150"]  
511715  2017-02-24 21:00:17                    110  
512430  2017-02-24 21:08:58     ["160", "80", "2"]  
513729  2017-02-24 21:35:45                    110  
515615  2017-02-24 22:21:30    ["80", "160", "80"]  

[519 rows x 7 columns]

Select two tasks:

In [146]:
two_tasks = data.loc[(data.task_id == 1901) | (data.task_id == 1902)]
two_tasks

task_id  user_id answer  confidence           start_time  \
4          1902     2827     75           2  2017-02-02 23:57:23   
6          1901     2827    150           4  2017-02-02 23:57:30   
2942       1901     2825    100           1  2017-02-03 03:32:16   
2989       1902     2825    200           1  2017-02-03 03:33:11   
3350       1902     2925     85           3  2017-02-03 03:42:09   
...         ...      ...    ...         ...                  ...   
512430     1901     4501     80           2  2017-02-24 21:08:50   
513729     1901     4416    110           4  2017-02-24 21:35:40   
513742     1902     4416    180           2  2017-02-24 21:35:51   
515615     1901     3011    150           1  2017-02-24 22:21:25   
515624     1902     3011      0           1  2017-02-24 22:21:40   

                   end_time                 cues  
4       2017-02-02 23:57:30    "Not enough data"  
6       2017-02-02 23:57:38    "Not enough data"  
2942    2017-02-03 03:32:21                  NaN  
2989    2017-02-03 03:33:16                  NaN  
3350    2017-02-03 03:42:18                  NaN  
...                     ...                  ...  
512430  2017-02-24 21:08:58   ["160", "80", "2"]  
513729  2017-02-24 21:35:45                  110  
513742  2017-02-24 21:35:56                  180  
515615  2017-02-24 22:21:30  ["80", "160", "80"]  
515624  2017-02-24 22:21:45  ["100", "250", "0"]  

[1040 rows x 7 columns]

only the indexs of the tasks:

In [147]:
two_tasks.index.values

array([     4,      6,   2942, ..., 513742, 515615, 515624], dtype=int64)

the index in the first place:

In [150]:
two_tasks.index.values[0]

4

can you find the maximum index? 

Select task_id = 1901 and answers = 70 (hint: use &)

In [ ]:
userX = 

now find the user_id of this user

In [152]:
u_idx = userX.index.values[0]
data.loc[u_idx,'user_id']

3509

# Statstics on data


Task #1901 is: "How many calories does 1 glass of milk contain?"
(I know that because I read the tasks. See [here](https://github.com/stanford-policylab/wisdom-of-crowds))
So let's look at that task. 

In [85]:
milk_calories = data[data.loc[:,'task_id'] == 1901]
milk_calories.head()

task_id  user_id answer  confidence           start_time  \
6        1901     2827    150           4  2017-02-02 23:57:30   
2942     1901     2825    100           1  2017-02-03 03:32:16   
3358     1901     2925    270           4  2017-02-03 03:42:18   
3764     1901     2841     90           2  2017-02-03 03:51:52   
4548     1901     2972    250           3  2017-02-03 19:44:55   

                 end_time               cues  
6     2017-02-02 23:57:38  "Not enough data"  
2942  2017-02-03 03:32:21                NaN  
3358  2017-02-03 03:42:28                NaN  
3764  2017-02-03 03:51:58                NaN  
4548  2017-02-03 19:45:01  "Not enough data"

side note: notice that the variable name is meaningful :-)

In [86]:
milk_calories.median()

task_id       1901.0
user_id       3730.0
answer         125.0
confidence       3.0
dtype: float64

In [87]:
milk_calories.mean()

task_id       1901.000000
user_id       3735.597303
confidence       2.643545
dtype: float64

This isn't working. There is no mean for the field 'answer'.
Maybe because the 'answer' is not numeric?

In [89]:
milk_calories.answer

6         150
2942      100
3358      270
3764       90
4548      250
         ... 
510135     80
511715    120
512430     80
513729    110
515615    150
Name: answer, Length: 519, dtype: object

It looks numeric. So maybe just ask the mean of the numeric only values?

In [90]:
milk_calories.mean(numeric_only= True)

task_id       1901.000000
user_id       3735.597303
confidence       2.643545
dtype: float64

That didn't work either!! Maybe data is not defined as numeric?

In [91]:
milk_calories.dtypes

task_id        int64
user_id        int64
answer        object
confidence     int64
start_time    object
end_time      object
cues          object
dtype: object

It's an object. Let's turn to numeric:

In [94]:

#milk_calories['answer'].astype(int)
pd.to_numeric(milk_calories.answer)
#milk_calories['answer'].astype(str).astype(int)

6         150.0
2942      100.0
3358      270.0
3764       90.0
4548      250.0
          ...  
510135     80.0
511715    120.0
512430     80.0
513729    110.0
515615    150.0
Name: answer, Length: 519, dtype: float64

Note that this does NOT change the varaible milk_calories.

In [95]:
milk_calories.dtypes

task_id        int64
user_id        int64
answer        object
confidence     int64
start_time    object
end_time      object
cues          object
dtype: object

---
wait.. what's the difference between dtypes and type()?

---


In [99]:
answer = pd.to_numeric(milk_calories['answer'])
answer.mean()

133.96686159844054

In [100]:
answer.median()

125.0

In [101]:
answer.mode() 

0    150.0
dtype: float64

What is more accurate in this case? mean, median or mode?
This is what google says (in the dataset the answer is 105):

<div>
<img src="milk.png" width="800"/>
</div>

<div>
<img src="sanity.png" width="400"/>
</div>

At this point, if you're not sure your answer is correct, it's a good idea to go and check the file. 


In [102]:
answer.describe()

count    513.000000
mean     133.966862
std       45.563281
min        2.000000
25%      100.000000
50%      125.000000
75%      150.000000
max      400.000000
Name: answer, dtype: float64

### categorical data:

In [103]:
milk_calories.cues

6             "Not enough data"
2942                        NaN
3358                        NaN
3764                        NaN
4548          "Not enough data"
                  ...          
510135    ["150", "166", "150"]
511715                      110
512430       ["160", "80", "2"]
513729                      110
515615      ["80", "160", "80"]
Name: cues, Length: 519, dtype: object

List of unique values

In [ ]:
milk_calories.cues.unique()

To see a list of unique values and how often they occur

In [ ]:
milk_calories.cues.value_counts()

# Treating missing values
---





Are there any missing values?

In [71]:
milk_calories.isnull().values.any()

True

Yes there are. How many?

In [72]:
milk_calories.answer.isnull().sum()

6

In [73]:
milk_calories.answer.count()

513

What is the percentage of nulls?

In [74]:
len(milk_calories.index)

519

So what is the percentage of null cells? Take it from here..

There is more then one way to do this..

In [75]:
milk_calories.answer

6         150
2942      100
3358      270
3764       90
4548      250
         ... 
510135     80
511715    120
512430     80
513729    110
515615    150
Name: answer, Length: 519, dtype: object

In [76]:
#This is one option (to the question percentage of null values)
#Note this is confusing since it looks like it is computing the mean, but it's not.
milk_calories.answer.notnull().mean()

0.9884393063583815

Was the mean computed correctly? What happened to the null values?
'nanmean' computes the mean ignoring the nulls

In [81]:
np.nanmean(pd.to_numeric(milk_calories.answer))


133.96686159844054

In [82]:
pd.to_numeric(milk_calories.answer).mean()

133.96686159844054

What can we do with the null values?

Remove these rows

What can we do with the null values?

Replace the nulls with the mean, median or mode

< CONTINUE> -the code is in the PDSH here: https://jakevdp.github.io/PythonDataScienceHandbook/03.04-missing-values.html

Remove the null values
Finally we get the mean:

What's the real answer? Which is better here? The median or the mean?

In [69]:
milk_calories['answer'].mode

<bound method Series.mode of 6         150
2942      100
3358      270
3764       90
4548      250
         ... 
510135     80
511715    120
512430     80
513729    110
515615    150
Name: answer, Length: 519, dtype: object>

Select all users with the last four digits of your id
TO COMPLETE
2827

How many users have id 2827?

for a task id that ends with the last three digits of your id (if there is no such task id, take two digits)
Check which is better, median or mean, 